File: crates/app-core/src/commands/issues.rs
//! Tauri commands for Issues — list/detail/create/edit/close/reopen/comment/
//! labels/assignees/milestones. Wraps [`ForgeProvider`] trait methods.
//!
//! Each command resolves the active forge provider via [`build_forge_provider`]
//! and dispatches to the corresponding trait method on a `spawn_blocking`
//! thread. Errors are stringified at the IPC boundary per Tauri convention.

use std::sync::Arc;

use forge_provider::{
    CreateIssueInput, EditIssuePatch, ForgeProvider, Issue, IssueDetail, IssueFilter, IssueState,
    Milestone,
};
use tauri::State;

use super::helpers::*;
use crate::state::AppState;

/// List issues for the current repo with optional filters.
///
/// The `state_filter` arg accepts `"open"`, `"closed"`, or `None` (=all).
#[tauri::command]
#[allow(clippy::too_many_arguments)]
pub async fn list_issues(
    state_filter: Option<String>,
    author: Option<String>,
    assignee: Option<String>,
    label: Option<String>,
    milestone: Option<u64>,
    text: Option<String>,
    limit: Option<u32>,
    state: State<'_, AppState>,
) -> Result<Vec<Issue>, String> {
    let provider: Arc<dyn ForgeProvider> = build_forge_provider(&state)?;
    let filter = IssueFilter {
        state: match state_filter.as_deref() {
            Some("open") => Some(IssueState::Open),
            Some("closed") => Some(IssueState::Closed),
            _ => None,
        },
        author,
        assignee,
        label,
        milestone,
        text,
    };
    let limit = limit.unwrap_or(50);
    run_blocking(move || {
        provider
            .list_issues(filter, limit)
            .map_err(|e| e.to_string())
    })
    .await
}

/// Fetch full detail (body + comments) for a single issue.
#[tauri::command]
pub async fn get_issue(number: u64, state: State<'_, AppState>) -> Result<IssueDetail, String> {
    let provider: Arc<dyn ForgeProvider> = build_forge_provider(&state)?;
    run_blocking(move || provider.get_issue(number).map_err(|e| e.to_string())).await
}

/// Create a new issue.
#[tauri::command]
pub async fn create_issue(
    title: String,
    body: String,
    labels: Vec<String>,
    assignees: Vec<String>,
    milestone: Option<u64>,
    state: State<'_, AppState>,
) -> Result<Issue, String> {
    let provider: Arc<dyn ForgeProvider> = build_forge_provider(&state)?;
    let input = CreateIssueInput {
        title,
        body,
        labels,
        assignees,
        milestone,
    };
    run_blocking(move || provider.create_issue(input).map_err(|e| e.to_string())).await
}

/// Edit an issue's title and/or body.
#[tauri::command]
pub async fn edit_issue(
    number: u64,
    title: Option<String>,
    body: Option<String>,
    state: State<'_, AppState>,
) -> Result<(), String> {
    let provider: Arc<dyn ForgeProvider> = build_forge_provider(&state)?;
    let patch = EditIssuePatch { title, body };
    run_blocking(move || {
        provider
            .edit_issue(number, patch)
            .map_err(|e| e.to_string())
    })
    .await
}

/// Close an open issue.
#[tauri::command]
pub async fn close_issue(number: u64, state: State<'_, AppState>) -> Result<(), String> {
    let provider: Arc<dyn ForgeProvider> = build_forge_provider(&state)?;
    run_blocking(move || provider.close_issue(number).map_err(|e| e.to_string())).await
}

/// Reopen a closed issue.
#[tauri::command]
pub async fn reopen_issue(number: u64, state: State<'_, AppState>) -> Result<(), String> {
    let provider: Arc<dyn ForgeProvider> = build_forge_provider(&state)?;
    run_blocking(move || provider.reopen_issue(number).map_err(|e| e.to_string())).await
}

/// Post a general comment on an issue.
#[tauri::command]
pub async fn add_issue_comment(
    number: u64,
    body: String,
    state: State<'_, AppState>,
) -> Result<(), String> {
    let provider: Arc<dyn ForgeProvider> = build_forge_provider(&state)?;
    run_blocking(move || {
        provider
            .add_issue_comment(number, &body)
            .map_err(|e| e.to_string())
    })
    .await
}

/// Add labels to an issue.
#[tauri::command]
pub async fn add_issue_labels(
    number: u64,
    labels: Vec<String>,
    state: State<'_, AppState>,
) -> Result<(), String> {
    let provider: Arc<dyn ForgeProvider> = build_forge_provider(&state)?;
    run_blocking(move || {
        provider
            .add_issue_labels(number, &labels)
            .map_err(|e| e.to_string())
    })
    .await
}

/// Remove labels from an issue.
#[tauri::command]
pub async fn remove_issue_labels(
    number: u64,
    labels: Vec<String>,
    state: State<'_, AppState>,
) -> Result<(), String> {
    let provider: Arc<dyn ForgeProvider> = build_forge_provider(&state)?;
    run_blocking(move || {
        provider
            .remove_issue_labels(number, &labels)
            .map_err(|e| e.to_string())
    })
    .await
}

/// Add assignees to an issue.
#[tauri::command]
pub async fn add_issue_assignees(
    number: u64,
    assignees: Vec<String>,
    state: State<'_, AppState>,
) -> Result<(), String> {
    let provider: Arc<dyn ForgeProvider> = build_forge_provider(&state)?;
    run_blocking(move || {
        provider
            .add_issue_assignees(number, &assignees)
            .map_err(|e| e.to_string())
    })
    .await
}

/// Remove assignees from an issue.
#[tauri::command]
pub async fn remove_issue_assignees(
    number: u64,
    assignees: Vec<String>,
    state: State<'_, AppState>,
) -> Result<(), String> {
    let provider: Arc<dyn ForgeProvider> = build_forge_provider(&state)?;
    run_blocking(move || {
        provider
            .remove_issue_assignees(number, &assignees)
            .map_err(|e| e.to_string())
    })
    .await
}

/// Set (or clear with `None`) the milestone on an issue.
#[tauri::command]
pub async fn set_issue_milestone(
    number: u64,
    milestone_id: Option<u64>,
    state: State<'_, AppState>,
) -> Result<(), String> {
    let provider: Arc<dyn ForgeProvider> = build_forge_provider(&state)?;
    run_blocking(move || {
        provider
            .set_issue_milestone(number, milestone_id)
            .map_err(|e| e.to_string())
    })
    .await
}

/// List all milestones for the current repo (for picker UIs).
#[tauri::command]
pub async fn list_milestones(state: State<'_, AppState>) -> Result<Vec<Milestone>, String> {
    let provider: Arc<dyn ForgeProvider> = build_forge_provider(&state)?;
    run_blocking(move || provider.list_milestones().map_err(|e| e.to_string())).await
}

File: src/lib/stores/issues.ts
/**
 * Issues store — manages issue list, detail, filters, labels & milestones cache.
 *
 * Mirrors the MR/PR store pattern. The labels and milestones caches are
 * lazily populated via `refreshLabelsCache()` / `refreshMilestonesCache()`
 * the first time a picker mounts.
 *
 * Migrated to the RepoState container (spec 08): the view state below is a thin
 * facade over the active repo's `IssuesSlice` (see `repo-state/IssuesSlice.ts`).
 */

import { derived, get } from "svelte/store";
import type { Issue, IssueDetail, IssueState, Label, Milestone } from "../types";
import * as api from "../api/tauri";
import { fetchIntoStore } from "../utils/store-helpers";
import { activeField, getActiveRepoState } from "./repo-state";

// The issue view state below is per-repo: it lives in the active repo's
// `IssuesSlice` so switching project tabs shows that repo's list, selection,
// filter, and picker caches. These facades proxy the active slice so existing
// component imports keep working.

/** Current state filter tab: open, closed, or all. */
export const issueStateFilter = activeField<IssueState | "all">((rs) => rs.issues.stateFilter);

/** Current list of issues matching the filter. */
export const issueList = activeField<Issue[]>((rs) => rs.issues.list);

/** Whether the list is currently loading. */
export const issueListLoading = activeField<boolean>((rs) => rs.issues.listLoading);

/** Currently selected issue number. */
export const selectedIssueNumber = activeField<number | null>((rs) => rs.issues.selectedNumber);

/** Full detail of the selected issue. */
export const issueDetail = activeField<IssueDetail | null>((rs) => rs.issues.detail);

/** Whether the detail view is loading. */
export const issueDetailLoading = activeField<boolean>((rs) => rs.issues.detailLoading);

/** Cache of repository labels (for pickers). Lazily loaded. */
export const labelsCache = activeField<Label[]>((rs) => rs.issues.labelsCache);
/** Whether the labels cache is currently loading. */
export const labelsCacheLoading = activeField<boolean>((rs) => rs.issues.labelsCacheLoading);

/** Cache of repository milestones (for pickers). Lazily loaded. */
export const milestonesCache = activeField<Milestone[]>((rs) => rs.issues.milestonesCache);
/** Whether the milestones cache is currently loading. */
export const milestonesCacheLoading = activeField<boolean>((rs) => rs.issues.milestonesCacheLoading);

/** Derived: `Map<number, Issue>` for cross-ref resolution by number. */
export const issueByNumber = derived(issueList, ($list) => {
  const map = new Map<number, Issue>();
  for (const i of $list) map.set(i.number, i);
  return map;
});

/** Fetch the issue list with the current state filter. */
export async function refreshIssueList(): Promise<void> {
  // Capture the target slice up front so a late response lands in the repo it
  // belongs to, even if the user switches tabs mid-flight (RepoState renders
  // only the active slice, so writing back into an inactive repo's slice is
  // correct and invisible).
  const slice = getActiveRepoState().issues;
  const filter = get(slice.stateFilter);
  const state = filter === "all" ? undefined : (filter as IssueState);
  slice.listLoading.set(true);
  try {
    const items = await api.listIssues(
      state,
      undefined,
      undefined,
      undefined,
      undefined,
      undefined,
      50,
    );
    slice.list.set(items);
    // Invalidate selection if the currently-selected issue vanished.
    const selected = get(slice.selectedNumber);
    if (
      selected !== null &&
      !items.some((item) => item.number === selected)
    ) {
      slice.selectedNumber.set(null);
    }
  } catch {
    slice.list.set([]);
    slice.selectedNumber.set(null);
  } finally {
    slice.listLoading.set(false);
  }
}

/** Load detail for a specific issue and mark it selected. */
export async function loadIssueDetail(number: number): Promise<void> {
  const slice = getActiveRepoState().issues;
  slice.selectedNumber.set(number);
  slice.detailLoading.set(true);
  try {
    const d = await api.getIssue(number);
    slice.detail.set(d);
  } catch {
    slice.detail.set(null);
  } finally {
    slice.detailLoading.set(false);
  }
}

/** Clear detail-only state without touching the list. */
export function clearIssueDetail(): void {
  getActiveRepoState().issues.clearDetail();
}

/** Reset all issue state (on project switch). */
export function clearIssueState(): void {
  getActiveRepoState().issues.clear();
}

/** Populate the labels cache (no-op on error — cache stays empty). */
export async function refreshLabelsCache(): Promise<void> {
  const slice = getActiveRepoState().issues;
  await fetchIntoStore(slice.labelsCache, slice.labelsCacheLoading, api.listLabels, []);
}

/** Populate the milestones cache (no-op on error — cache stays empty). */
export async function refreshMilestonesCache(): Promise<void> {
  const slice = getActiveRepoState().issues;
  await fetchIntoStore(
    slice.milestonesCache,
    slice.milestonesCacheLoading,
    api.listMilestones,
    [],
  );
}

// ─── Write operations ────────────────────────────────────────────────

/** Create a new issue and refresh the list. */
export async function createIssue(
  title: string,
  body: string,
  labels: string[],
  assignees: string[],
  milestone: number | null,
): Promise<Issue> {
  const result = await api.createIssue(title, body, labels, assignees, milestone);
  await refreshIssueList();
  return result;
}

/** Edit an issue and reload detail. */
export async function editIssue(
  number: number,
  title?: string,
  body?: string,
): Promise<void> {
  await api.editIssue(number, title, body);
  await loadIssueDetail(number);
}

/** Close an issue and refresh list + detail. */
export async function closeIssue(number: number): Promise<void> {
  await api.closeIssue(number);
  await loadIssueDetail(number);
  await refreshIssueList();
}

/** Reopen an issue and refresh list + detail. */
export async function reopenIssue(number: number): Promise<void> {
  await api.reopenIssue(number);
  await loadIssueDetail(number);
  await refreshIssueList();
}

/** Post a comment on an issue and refresh detail. */
export async function addIssueComment(
  number: number,
  body: string,
): Promise<void> {
  await api.addIssueComment(number, body);
  await loadIssueDetail(number);
}

/** Add labels and reload detail. List row counts tolerate staleness until next refresh. */
export async function addIssueLabels(
  number: number,
  labels: string[],
): Promise<void> {
  await api.addIssueLabels(number, labels);
  await loadIssueDetail(number);
}

/** Remove labels and reload detail. List row counts tolerate staleness until next refresh. */
export async function removeIssueLabels(
  number: number,
  labels: string[],
): Promise<void> {
  await api.removeIssueLabels(number, labels);
  await loadIssueDetail(number);
}

/** Add assignees and reload detail. List row counts tolerate staleness until next refresh. */
export async function addIssueAssignees(
  number: number,
  assignees: string[],
): Promise<void> {
  await api.addIssueAssignees(number, assignees);
  await loadIssueDetail(number);
}

/** Remove assignees and reload detail. List row counts tolerate staleness until next refresh. */
export async function removeIssueAssignees(
  number: number,
  assignees: string[],
): Promise<void> {
  await api.removeIssueAssignees(number, assignees);
  await loadIssueDetail(number);
}

/** Set (or clear) the milestone on an issue and reload detail. List row counts tolerate staleness until next refresh. */
export async function setIssueMilestone(
  number: number,
  milestoneId: number | null,
): Promise<void> {
  await api.setIssueMilestone(number, milestoneId);
  await loadIssueDetail(number);
}
